fairseq/trainer.py

```diff
@@ -342,10 +344,17 @@ def maybe_no_sync():
                 else:
                     raise e
 
+            if self.xla and i < len(samples)-1:
+                # tpu-comment: every xla operation before marking step is
+                # appended to the IR graph, and processing too many batches
+                # before marking step can lead to OOM errors.
+                # To handle gradient accumulation use case, we explicitly
+                # mark step here for every forward pass without a backward pass
+                xm.mark_step()
```

Comment:
is the sample iteration above inside of the ParallelLoader one?
Because the ParallelLoader already barrier.

Reply:
parallel loader yields `samples`. with `--update-freq=10`, it will yield 10 batches per step. I tried to list out how barriers would look like in point 4 inside https://github.com/pytorch-tpu/fairseq/pull/22#issuecomment-557223044